### PR TITLE
Add a systray applet

### LINKED ISSUE #146: Make a proper systray applet for Arch-Update
This has been requested quite a few times as it allows an easy integration within the system regardless of the panel/bar used by any DE/WM (as long as it has systray support, obviously).  

It is also way more elegant when it comes to dynamic icon changes (e.g. when there are pending available updates).
Indeed, the way this "dynamic icon changes" has been initially implemented on Arch-Update's side is *hacky* and do not even work properly for most people (see [this comment](https://github.com/Antiz96/arch-update/pull/145#issuecomment-2096960519)).

I'm still in favor of keeping the `.desktop` file anyway (for people that don't have/want a systray) but only using one single "regular/non-changing" icon (as most other applications do).

When it comes to the systray applet specs, here's what I could think of so far:
- Ideally written in a high-level, intelligible and widely used programing language (for instance Python or Rust).
- Can be launched by a systemd-service (or by running a simple command, such as `arch-update --applet` or `arch-update-applet`, that one can put in their autostart apps or WM config) to have it running at startup.
- For icons:
    - After a second thought, I'm not sure how relevant some icons are (e.g. the 'checking' and 'installing' icons) and how hard handling them in such an applet would be. So, if anything, we can eventually drop them and only have an "up to date" and an "update available" icons.
    - Depending on how hard it is to handle, having the applet showing the number of pending updates (such as what the [Arch Updates Indicator Gnome extension](https://extensions.gnome.org/extension/1010/archlinux-updates-indicator/) is doing) would be a nice plus!

To be honest, I don't have a precise idea on how to develop such an applet for now. I'm gonna start digging into it as soon as I'll be able to but, obviously, any help would be greatly appreciated! :smile:

### PR DESCRIPTION
**Pull request by trigg**
- One-file python script added as `src/script/arch-update-tray.py`
- Installs to `/usr/bin/arch-update-tray` by default
- Clicking icon will launch the updater
- Launching the updater is done via the `.desktop` file, so glib dependency is required to launch it
- Statefile is monitored and if the first line of the file starts with 'arch-update' the entire first line, stripped of whitespace, is used as the icon.

# Dependencies

- `python3`
- `python-pyqt6`
- `qt6-svg`
- `glib2`

Closes #146

### CODE REVIEW

**Comment by trigg** [comment]
A note: If we want to pick sides (most likely for the sake of simplicity) between the toolkits, this could be cut down noticably.

I simply wanted to try my hand at making it as available as I could.

**Comment by Antiz96** [comment]
Just some quick comments after a first read:

> [...]
>     * * Due to a fun quirk in GTK AppIndicator, clicking the icon opens a menu with the update option in it.

Uh... Too bad :(

> [...]
>     * Launching the updater is done via the `.desktop` file, so another dependency is required to launch it (more about this later)

Is there a specific reason to launch it via the `.desktop` file rather than calling the `arch-update` bin directly?

> [...]
>     * environment variable `ARCH_UPDATE_TK` is checked. If set to `gtk3` or `qt6` then prefer that toolkit, if multiple toolkits are installed. Defaults to `qt6` for no reason at all.

"Defaults to `qt6` for no reason at all." <-- Haha. Well, if the QT6 toolkit allows a direct launch of the updater (where the gtk one opens a menu first), that's a good enough reason I guess :D

> [...]
> 
> # Dependencies
> 
> Naturally this is a _Lot_ of extra dependencies above what the project requires previously. I've tried to lessen the impact by allowing alternatives.
> 
>     * `python3`

Fair enough :rofl: 

> [...]
> Note that I'm not certain the list here is comprehensive, testing will be necessary

Alright, no problem. Should be able to test that in a minimal environment to make sure nothing's missing :)

> [...]
> A note: If we want to pick sides (most likely for the sake of simplicity) between the toolkits, this could be cut down noticably.
> I simply wanted to try my hand at making it as available as I could.

Well, I'm in a favor of making this as simple as possible (at least to start things of), ideally both in terms of dependencies and about the python script itself (also for maintenance reasons. While the script looks fairly simple and intelligible, I'm definitely not as comfortable with Python than I am with bash :sweat_smile:). So I'm all in favor of picking a side on that front.

Usually I would personally choose GTK over QT for some reasons, but if the "menu quirk" cannot be fixed or worked around, I'm afraid that's a non-starter :/
In such case, I guess going with the QT toolkit and, hopefully, **not** the `kde-cli-tools` launcher would do :stuck_out_tongue_winking_eye: 

Once again, thank you so much for your incredible work on this! :pray: 
I'll take some time (hopefully this evening) to create the missing PR, merge the other ones, and start testing this Python script! :snake:

**Comment by Antiz96** [comment]
I also saw you made `pylint` exclusions/references in the script. Can I had `pylint` to the test CI?

**Comment by trigg** [comment]
> Is there a specific reason to launch it via the .desktop file rather than calling the arch-update bin directly?

It nets your the users preferred terminal. Running it directly will be unhelpful as it'll get no terminal, and we *could* take guesses at which terminal installed the user would prefer.

The launchers used will do it in a more sensible way than I could manage in a handful of lines of code

**Comment by Antiz96** [comment]
> > Is there a specific reason to launch it via the .desktop file rather than calling the arch-update bin directly?
> 
> It nets your the users preferred terminal. Running it directly will be unhelpful as it'll get no terminal, and we _could_ take guesses at which terminal installed the user would prefer.
> 
> The launchers used will do it in a more sensible way than I could manage in a handful of lines of code

Oh yeah right, good catch!

**Comment by trigg** [comment]
>  "menu quirk" cannot be fixed or worked around, I'm afraid that's a non-starter :/

Yeah, this is a huge problem for me. From everything I'm reading AppIndicator is dropped entirely in Gtk4, and they put as much effort into breaking it as possible before that happened too.

I keep thinking there should be a toolkit-agnostic way to do this, but I've not found it yet

**Comment by Antiz96** [comment]
> Yeah, this is a huge problem for me. From everything I'm reading AppIndicator is dropped entirely in Gtk4, and they put as much effort into breaking it as possible before that happened too.
> 

Well, I guess that coincides with the fact that Gnome wants to get rid of systrays altogether :P 

I guess this is also an additional legit reason to go with QT?

> I keep thinking there should be a toolkit-agnostic way to do this, but I've not found it yet

Hopefully there is, maybe we'll find one at some point ^^

**Comment by Antiz96** [comment]
I added pylint to CI. Can I let you check [the output](https://github.com/Antiz96/arch-update/actions/runs/8991431804/job/24699028107?pr=149) to see what can be addressed (or eventually ignored if needed)?

I assume that picking a side regarding the toolkit will allow to remove some parts of the code. Regarding the apparent GTK position regarding AppIndicator, I guess the safe choice is to go with QT6.
Glib2 feels like a good choice to me regarding the launcher.

**Comment by trigg** [comment]
`C0103: Module name "arch-update-tray" doesn't conform to snake_case naming style (invalid-name)`

My opinion is this is one for "we don't care"... the import ones we couldn't get around without switching the project to using pip or some other python-first build system, but this one is "we want your file to use underscores and not minus" ... which I don' think matters here as this isn't a module to begin with

**Comment by Antiz96** [comment]
> `C0103: Module name "arch-update-tray" doesn't conform to snake_case naming style (invalid-name)`
> 
> My opinion is this is one for "we don't care"... the import ones we couldn't get around without switching the project to using pip or some other python-first build system, but this one is "we want your file to use underscores and not minus" ... which I don' think matters here as this isn't a module to begin with

Yeah fair enough, sounds good to me. :+1: 
Seems like CI is now green :heavy_check_mark: :tada:

**Comment by trigg** [comment]
Do we want it stripped of GTK before going ahead?

It brings the least with it and has the higher bar for requirements

**Comment by Antiz96** [comment]
Well, I was about to ask your opinion about going with python-pyqt6 + glib2 but it seems your mind is set :slightly_smiling_face: 
Let's strip it then!

**Comment by trigg** [comment]
Something I didn't mention earlier...

Python pulls in the gettext translations set by the rest of this package, but never ends up using it in any meaningful way. For now I'll remove it again, but it's always an option to have menus etc with translated text

**Comment by Antiz96** [comment]
> A bit confusing choosing glib2 and pyqt6 but...

Why is that? To give you my train of thoughts: pyqt6 is the only option listed to go with the QT6 toolkit. As for glib2, it has the advantage of being a dependency for *a lot* of packages, meaning it is probably already installed on basically every people's system. But am I missing something? 
Is this because it's intended to get along with `kde-cli-tools`? (See below)

> ...kioclient automatically assumes konsole in my case.

I don't mind QT but I'd like to avoid KDE stuff if possible. I'm genuinely triggered by the such "assumptions" as you described above + the ~~insanely~~ huge amount of dependencies they require. :spiral_calendar:  
 
> It'll make requirements page much simpler than trying everything-and-kitchen-sink.

That's the aim here. Keep it simple as they say :)

**Comment by Antiz96** [comment]
> Something I didn't mention earlier...
> 
> Python pulls in the gettext translations set by the rest of this package, but never ends up using it in any meaningful way. For now I'll remove it again, but it's always an option to have menus etc with translated text

Alright, thanks for info. I don't intend to add any kind of menu for now so it's fine dropping it.
But good to know still.

**Comment by trigg** [comment]
> As for glib2, it has the advantage of being a dependency for a lot of packages

Oh no doubt, but glib being part of the gtk ecosystem it's ... less than typical to mix both in one application.

In our case I believe it's the best bet, no doubts.

> by the such "assumptions" as you described

Yeah, I'm not a KDE user, and I've used QT in projects before but always shy away from KDE (also Gnome)

> People installing with the default prefix 

Ah woops, Yeah I'm on it

File: src/script/arch-update-tray.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    STATE_FILE = os.path.join(
+        os.environ['HOME'], '.local', 'state', 'arch-update', 'current_state')
+if not os.path.isfile(STATE_FILE):
+    log.error("Statefile does not exist: %s", STATE_FILE)
+    sys.exit(1)
+
+
+def arch_update():
+    """ Launch with terminal """
+    update = "/usr/share/applications/arch-update.desktop"
```
[the remark this answers is not in the document]
Comment (Antiz96) [reply src/script/arch-update-tray.py:28]: https://github.com/Antiz96/arch-update/pull/149/commits/b2b3e4dde1e76c1fd1545c763e2c35f40f71887e

**Comment by Antiz96** [comment]
I'm gonna start adding the remaining specs for the systray to this PR (simple arg to launch it + systemd service) as well as adapting documentation accordingly. I'll start testing the systray itself tomorrow.

Any thoughts on the below? 
> Depending on how hard it is to handle, having the applet showing the number of pending updates (such as what the [Arch Updates Indicator Gnome extension](https://extensions.gnome.org/extension/1010/archlinux-updates-indicator/) is doing) would be a nice plus!

Retrieving the number of pending updates during checks and adding it to the "current_state" file (or another file) would be trivial but I'm not sure how hard it would be to integrate it to the systray.
If it represents too much effort or it is too hard or whatever, let's just forget it. If anything, we can think about it in a later time or simply drop it. No big deal :)

**Comment by Antiz96** [comment]
ToDo reminder:
- Add systray deps in the READMEs
- Replace the "icon changing" parts by the systray applet in documentation (READMEs and man pages)
- Remake the screenshots in the READMEs ( :sob: )
- Update po files

**Comment by trigg** [comment]
> Getting the number of pending updates during checks and adding it to the "current_state" file (or another file) would be trivial but I'm not sure how hard it would be to integrate it to the systray.

I honestly don't know how hard or easy this will be. I'm AFK for most of today but looking into the docs and spec is my next task.

**Comment by Antiz96** [comment]
> I honestly don't know how hard or easy this will be. I'm AFK for most of today but looking into the docs and spec is my next task.

Sure, no problem! As I said, if it requires too much searching or any kind of burdens in the implementation, let's just forget it. That would be a plus but definitely not mandatory :)

**Comment by Antiz96** [comment]
I made some tests with the systray applet and found two issues:

1 - No icons are shown, the python script reports: `QSystemTrayIcon::setVisible: No Icon set` (tried both with the `/usr/local` and `/usr` prefixes when installing). I can confirm the icons are correctly installed and works well with `notify-send` for instance though.
2 - Launching `arch-update` requires a double-click. Ideally, a single click should be enough.

**Comment by Antiz96** [comment]
According to the `QSystemTrayIcon` documentation, I expected the following change to solve the single-click but apparently it does not:
```
diff --git a/src/script/arch-update-tray.py b/src/script/arch-update-tray.py
index ec698cc..47cc08a 100755
--- a/src/script/arch-update-tray.py
+++ b/src/script/arch-update-tray.py
@@ -50,9 +50,10 @@ class ArchUpdateQt6:
             icon = QIcon.fromTheme(contents)
             self.tray.setIcon(icon)

-    def update(self):
+    def update(self, reason):
         """ Start arch-update """
-        arch_update()
+        if reason == QSystemTrayIcon.ActivationReason.Trigger:
+           arch_update()

     def __init__(self, statefile):
         """ Start Qt6 System Tray """
```

**Comment by trigg** [comment]
> 2 - Launching `arch-update` requires a double-click. Ideally, a single click should be enough.

A single click is all I used on wf-panel. Shouldn't be different elsewhere but I'll do testing tonight

**Comment by Antiz96** [comment]
> I made some tests with the systray applet and found two issues:
> 
> 1 - No icons are shown, the python script reports: `QSystemTrayIcon::setVisible: No Icon set` (tried both with the `/usr/local` and `/usr` prefixes when installing). I can confirm the icons are correctly installed and works well with `notify-send` for instance though. 2 - Launching `arch-update` requires a double-click. Ideally, a single click should be enough.

**Comment by Antiz96** [comment]
Missclick :sweat:

**Comment by Antiz96** [comment]
> > 2 - Launching `arch-update` requires a double-click. Ideally, a single click should be enough.
> 
> A single click is all I used on wf-panel. Shouldn't be different elsewhere but I'll do testing tonight

Woops, nevermind... It seems like this issue was linked to my initial *borked* test environment. Sorry for the false negative...
However, I was able to reproduce the `QSystemTrayIcon::setVisible: No Icon set` in multiple environments. I'm digging into it.

**Comment by Antiz96** [comment]
> However, I was able to reproduce the `QSystemTrayIcon::setVisible: No Icon set` in multiple environments. I'm digging into it.

Installing `qt6-svg` solves that. I'm adding it to the dependency list.

**Comment by Antiz96** [comment]
Alright, I think we're good for the systray applet itself, everything seems to work as intended. :partying_face: 
Once again, thanks a lot for your incredible work and your precious help on that front @trigg! :pray: 

I'll proceed with the remaining parts listed [here](https://github.com/Antiz96/arch-update/pull/149#issuecomment-2099465622) as soon as possible. Once done, and after an additional round of extended tests (just for good measures), we can merge this PR! :smile: 

I should have the new icons set merged by then so we will hopefully be able to cut the `v2.0.0` release soon after :tada:

**Comment by trigg** [comment]
Looks good! Glad we found the rest of the dependencies, I had a feeling something else was needed

**Comment by Antiz96** [comment]
@trigg Hey! Can you sync/update your fork so I can cherry-pick the [new icons set](https://github.com/Antiz96/arch-update/pull/152) into that PR please? :)

I'm trying to get this merged today \o/

**Comment by trigg** [comment]
Done that now

**Comment by Antiz96** [comment]
Thx!

**Review by Antiz96 — APPROVED** [review]
It took quite some time but I finally adapted documentation, added the few remaining specs and ran tests. I can finally say it: LGTM!

One last time, thanks for your incredible work and help on this.

I'll tag a new release shortly. Let's ship this applet to people! :rocket:

**Pull request merged**